src/lib.rs

```diff
@@ -0,0 +1,222 @@
+                    .to_string_lossy()
+                    .into_owned(),
+            };
+            let func_name = match code.co_name.is_null() {
+                true => "null".to_string(),
+                false => CStr::from_ptr(PyUnicode_AsUTF8(code.co_name))
+                    .to_string_lossy()
+                    .into_owned(),
+            };
+            Ok(FrameData {
+                func_name,
+                file_name,
+                identifier: frame as usize,
+                yielded_coroutine,
+            })
+        }
+    }
+}
+
+/// end can be None due to lack of Return callback
+#[cfg(Py_3_8)]
+struct ProfilerEntry {
+    file_name: String,
+    func_name: String,
+    start: Instant,
+    end: Option<Instant>,
+}
+
+/// Profiler context to be used as a value for ContextVar.
+#[pyclass]
+struct ProfilerContext {
+    entries: std::cell::Cell<HashMap<usize, ProfilerEntry>>,
+}
+
+/// Format an entry into file_name, func_name and duration.
+fn format_entry(entry: &ProfilerEntry) -> (String, String, Option<u128>) {
+    let duration = match entry.end {
+        Some(v) => Some(v.duration_since(entry.start).as_nanos()),
+        None => None,
+    };
+    (entry.file_name.clone(), entry.func_name.clone(), duration)
+}
+
+#[pymethods]
+impl ProfilerContext {
+    #[new]
+    fn new() -> Self {
+        ProfilerContext {
+            entries: std::cell::Cell::new(HashMap::new()),
+        }
+    }
+    #[getter]
+    fn entries(&mut self) -> PyResult<Vec<(String, String, Option<u128>)>> {
+        let mut result = Vec::new();
+        for (_, entry) in self.entries.get_mut().iter() {
+            result.push(format_entry(entry));
+        }
+        Ok(result)
+    }
+}
+
+fn get_context(py: Python, obj: *mut pyo3::ffi::PyObject) -> Option<Py<ProfilerContext>> {
+    let mut context_obj: *mut pyo3::ffi::PyObject = std::ptr::null_mut();
+
+    unsafe {
+        match PyContextVar_Get(obj, std::ptr::null_mut(), &mut context_obj) {
+            0 => (),
+            _ => return None,
+        };
+    }
+
+    unsafe { Py::from_owned_ptr_or_opt(py, context_obj) }
+}
+
+/// Our profiler callback
+extern "C" fn callback(
+    obj: *mut pyo3::ffi::PyObject,
+    frame: *mut PyFrameObject,
+    what: c_int,
+    _arg: *mut pyo3::ffi::PyObject,
+) -> c_int {
+    let event: TraceEvent = match what.try_into() {
+        Ok(event) => match event {
+            TraceEvent::Call | TraceEvent::Return => event,
+            _ => return 0,
+        },
+        _ => return 0,
+    };
+
+    let frame_data = match FrameData::try_from(frame) {
+        Ok(v) => v,
+        Err(err) => {
+            println!("{}", err);
+            return 0;
+        }
+    };
+
+    let gil = Python::acquire_gil();
+    let py = gil.python();
+
+    let context = match get_context(py, obj) {
+        Some(v) => v,
+        _ => return 0,
+    };
+    let mut context = context.as_ref(py).borrow_mut();
```

Comment:
Consider something like:
```suggestion
    let mut context = get_context(py, obj)
        .and_then(|v| Some(v.as_ref(py).borrow_mut()))
        .or(0);
```

Reply:
It didn't seem to work - I think or(0) doesn't return 0 in the main function.. which is what I need :(

Reply:
Oh right, `or` takes (and returns) an `Option`, so here you would need `unwrap_or(0)` instead.
Actually going back over [`Option`'s docs](https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.map_or) I think it can also be written as
```suggestion
    let mut context = get_context(py, obj).map_or(0, |v| v.as_ref(py).borrow_mut());
```
that's quite concise :)